enhancements/accelerators/out-of-tree-drivers.md

```diff
@@ -0,0 +1,539 @@
+
+This way we ensure that an image is pulled and placed only on the node where the kernel matches. It is the responsibility of the DriverContainer image builder to name the image the correct way.
+
+If an administrator updates the CR with a new driver version will taint the node rebuild the DriverContainer and un-taint the node with the new drivers running on a node. 
+
+#### Updates in OpenShift
+
+Updates in OpenShift can happen in two ways: 
+1. Only the payload (operators and needed parts on top of the OS) is updated
+2. The payload and the OS are simultaneously updated. 
+
+The first case is "easy" the new version of the operator will reconcile the expected state and verify that all parts of the special resource stack are working and then "do" nothing. 
+
+For the second case, the new operator will reconcile the expected state and see that there is a mismatch regarding the kernel version of the DriverContainer and the updated Node. It will try to pull the new image with the correct kernel version. If the correct DriverContainer cannot be pulled, will update the BuildConfig with the right kernel version and OpenShift will reinitiate the build since we have the ConfigChange trigger as described above. 
+
+Besides the ConfigChange trigger, we also added the ImageChange trigger, which is important when the base image is updated due to CVE or other bug fixes. For this to happen automatically we are leveraging ImageStreams of OpenShift, an ImageStream is a collection of tags that gets automatically updated with the latest tags. It is like a container repository that represents a virtual view of related images.
+
+To be always up to date another possibility would be to register a github/gitlab webhook so every time the DriverContainer code changes a new container could be built. One has just to make sure that the webhook is triggered on a specific release branch, it is not advisable to monitor a fast moving branch (e.g. master) that would trigger frequent builds. 
+
+#### Special Resource Driver Downgrade
+
+Having a look at the example CRs above we can see that one can provide a driver version for a specific hardware aka DriverContainer. will take care of updating the BuildConfig and DriverContainer manifests. Tainting the node as **specialresource.openshift.io/downgrade=true:NoExecute** will evade all running Pods and the DriverContainer can be restarted. When the DriverContainer is again up and running, the node can be un-tainted to allow workloads to be scheduled on the node again. 
+ 
+ 
+#### Update proactive DriverContainers
+
+A preferable workflow for updates could also be to be proactive on updates. When OpenShift is updated we would need a mechanism (notification, hook for updates) to provide the kernel version of the next update to before attempting the upgrade and rebooting the nodes (needinfo installer/CVO team) .
```

Comment:
I'm a bit fuzzy on how machines decide which RHCOS to pivot too.  But maybe your operator could keep an eye on that target, and drop PDB guards onto the DriverContainers to protect them from PDB drains if it saw a target RHCOS with an incompatible kernel?

Reply:
I was thinking of something more "drastic" measure. Actually I want to prevent any action happening before SRO can give green lights that the driver will work on the next version. [https://github.com/openshift/enhancements/pull/357#issuecomment-748112774](https://github.com/openshift/enhancements/pull/357#issuecomment-748112774)